### DISCUSSION TITLE
Routing HL7 messages based on value in the message

### DISCUSSION THREAD

**Discussion by DaanOttinkTopicus** [nextgenhealthcare/connect#5162] [Q&A] [answered]
Hi all, 

Looking for some advise on how to improve our current setup, which is as follows:

The first channel is the one who receives a HL7 v2.5 message as it's source (TCP receiver) and sends it to its two destinations
1. a third party (TCP sender)
2. Another channel (Channel writer)

The second channel is a channel reader which reads the input send by the first channel. This channel has close to 50 destinations, all of these have a filter. The filter checks the ORC.4.1 segment from the HL7 message and based on the value either filters the message or sends it through. 

This approach works fine but has a couple of drawbacks:
1. A change to destination 35 requires a restart of the channel and thus a restart of all 50 destinations
2. Every message gets copied 50 times but only send once, resulting in a fast growing database. Messages older than x days get pruned but there are still a great number of duplicates stored. 

Is there any way to route the incoming messages directly to a channel based on the value in the ORC.4.1 without duplicating the message to 50 destinations?

Thanks.

**Comment by pacmano1** [comment, marked as answer]
You need to be using destination set filtering to prevent the message copy.   Destination set filtering allows you to remove all, one, or several destinations.   That decision is made in a source transformer or filter.   

<img width="1118" alt="image" src="https://user-images.githubusercontent.com/44065187/165761222-1d247389-2a1f-408a-a699-01704c19687a.png">

e.g.
```JS
destinationSet.removeAllExcept([3]);
```
I would probably do your set up on one channel.   You don't need the second one.

**Reply by pacmano1** [reply]
As for "A change to destination 35 requires a restart", what kind of changes are you making?   If they are port / ip changes,  those could be sourced at run time via maps.   As could other potential message logic.

**Reply by DaanOttinkTopicus** [reply]
I was not aware of this feature, maybe because we're still on an older version. We are currently in the process of upgrading to 4.0.1, I'm going to look into it. Thanks 😄 

EDIT: Yes, changes are often made to ip/port or other settings.

**Reply by jonbartels** [reply]
When calling `destinationSet.removeAllExcept([3]);` or similar `remove` methods ALWAYS check the return value.

```
var removed = destinationSet.removeAllExcept("Whoops I typo'd the destination name!");
if(!removed){
 throw "The developer made a mistake and no destinations were removed. Check that the destination name or ID is correct!"
}
```

**Reply by pacmano1** [reply]
Developers never make mistakes.

**Reply by pacmano1** [reply]
Destination set filtering was added in Mirth 3.1.   https://github.com/nextgenhealthcare/connect/wiki/3.5.0---What's-New#destination-set-filtering-improvements mentions that in the context of an update in 3.5  they lets you use that function without writing Javascript.

**Reply by DaanOttinkTopicus** [reply]
I wasn't aware of destination set filtering, apparently not a new feature 😅  👍
